util-print.c

```diff
@@ -242,23 +242,22 @@ ts_frac_print(netdissect_options *ndo, long usec)
  *   if date_flag == WITH_DATE print YY:MM:DD before HH:MM:SS.FRAC
  */
 static void
-ts_date_hmsfrac_print(netdissect_options *ndo, long sec, long usec,
+ts_date_hmsfrac_print(netdissect_options *ndo, const struct timeval *tv,
 		      enum date_flag date_flag, enum time_flag time_flag)
 {
-	time_t Time = sec;
 	struct tm *tm;
 	char timebuf[32];
 	const char *timestr;
 
-	if ((unsigned)sec & 0x80000000) {
+	if ((unsigned)tv->tv_sec & 0x80000000) {
```

Comment:
Presumably the reason why we decided to do the test that way was to handle systems with an unsigned `time_t`.  We'll need to figure out how to make that work.

Reply:
I propose to rebase and merge this PR as it is and to fix this other subject  after.

Reply:
More information here:
https://github.com/the-tcpdump-group/tcpdump/commit/cc190c1b5846bd221cec2e584759637ef7a27ffc#commitcomment-30024589

Reply:
I cannot remember why my patch was not applied.
https://github.com/the-tcpdump-group/tcpdump/files/2280041/time2.patch.zip

Reply:
Because it treat 'sec' as unsigned and remove the 
```
	if ((unsigned)sec & 0x80000000) {
		ND_PRINT("[Error converting time]");
		return;
        }
```

Reply:
> I cannot remember why my patch was not applied.
> https://github.com/the-tcpdump-group/tcpdump/files/2280041/time2.patch.zip

Was that part of a pull request, or an attachment to an issue?

> ```
> diff --git a/util-print.c b/util-print.c
> index 45da51f7..0996fe35 100644
> --- a/util-print.c
> +++ b/util-print.c
> @@ -266,15 +266,10 @@ static void
>  ts_date_hmsfrac_print(netdissect_options *ndo, int sec, int usec,
>  		      enum date_flag date_flag, enum time_flag time_flag)
>  {
> -	time_t Time = sec;
> +	time_t Time = (unsigned)sec;
>  	struct tm *tm;
>  	char timestr[32];
>  
> -	if ((unsigned)sec & 0x80000000) {
> -		ND_PRINT("[Error converting time]");
> -		return;
> -	}
> -
>  	if (time_flag == LOCAL_TIME)
>  		tm = localtime(&Time);
>  	else
> ```

`time_t Time = (unsigned)sec;` is OK until 2106.  As long as tcpdump is build on Windows with a 64-bit `time_t`, it will handle times between 2038 and 2106.

Reply:
> I propose to rebase and merge this PR as it is and to fix this other subject after.

We should probably at least leave behind comments indicating places where there are Y2.106K issues.

Reply:
> Was that part of a pull request, or an attachment to an issue?

(attachment) It was from: https://github.com/the-tcpdump-group/tcpdump/commit/cc190c1b5846bd221cec2e584759637ef7a27ffc#commitcomment-30042493.

Reply:
A second commit is needed for the part outside the warning fixes (unsigned in pure 32-bit), etc..
Before doing these changes, should we use `tttt` for tcpdump in `TESTrun` to show the year?

Reply:
> Before doing these changes, should we use tttt for tcpdump in TESTrun to show the year?

Yes, to catch more issues with processing timestamps.

Reply:
Doing a `(unsigned)tv->tv_sec & 0x80000000` mask on a seconds time_t value that can be 64 bits is incorrect.
Must be removed or updated.

Reply:
Even if *we* consider tv_sec as unsigned in the *pure* 32-bit case (without glibc _TIME_BITS=64), use `localtime(&tv->tv_sec)` or `gmtime(&tv->tv_sec)`  make it consider as signed by these functions, thus Y2038 problem remains.

Reply:
The only reason that comes to mind for testing whether the 0x80000000 bit of the `time_t` is set, rather than testing whether the `time_t` is negative, would be to avoid warnings if `time_t` is unsigned.

However, if it's unsigned, presumably it'd handle values > 2^31-1, so that's not an issue.

I don't know what platforms, if any, have an unsigned `time_t`; my inclination, at this point, is to test for the `time_t` value being < 0, and, if that gets a "test is always false" warning on any platform, just `#ifdef` out the test on that platform.

Reply:
> my inclination, at this point, is to test for the `time_t` value being < 0

Yes, done in the second commit.